gh-141004: Document missing `PyCFunction*` and `PyCMethod*` APIs (#141253)
python/cpython: Doc/c-api/structures.rst

   slot.  This is helpful because calls to PyCFunctions are optimized more
   than wrapper object calls.

.. c:var:: PyTypeObject PyCMethod_Type

   The type object corresponding to Python C method objects. This is
   available as :class:`types.BuiltinMethodType` in the Python layer.

.. c:function:: int PyCMethod_Check(PyObject *f)

>    Return true if *f* is an instance of the :c:type:`PyCMethod_Type` type

Reviewer: ```suggestion
   Returns ``true`` if *f* is an instance of the :c:type:`PyCMethod_Type` type. Otherwise returns ``false``.
```

Reviewer 2: I do not think it is necessary. We do not do so consistently.

Author: We use the current style throughout most of the documentation. See [`PyFunction_Check`](https://docs.python.org/3/c-api/function.html#c.PyFunction_Check), for example:

> Return true if o is a function object (has type [PyFunction_Type](https://docs.python.org/3/c-api/function.html#c.PyFunction_Type)).

   .. versionadded:: 3.9

.. c:var:: PyTypeObject PyCFunction_Type

   The type object corresponding to Python C function objects. This is
   available as :class:`types.BuiltinFunctionType` in the Python layer.

> .. c:function:: int PyCFunction_Check(PyObject *f)

Reviewer 2: Keep a consistent name for the parameter. For `_Check` functions, I would prefer having `op` everywhere.

>    Return true if *f* is an instance of the :c:type:`PyCFunction_Type` type

Reviewer 2: Out of curiosity, is it the standard formulation?

Author: What do you mean?

Reviewer 2: I'm surprised by the "instance of the X type". I would have just said "instance of X" but if we use the same sentence everywhere else, I'd prefer consistency.

Author: We use both forms as far as I can tell.

Reviewer 2: Let's keep this then.

>    or a subtype of it. This function always succeeds.

Reviewer: ```suggestion
   Returns ``true`` if *f* is an instance of the :c:type:`PyCFunction_Type` type
   or a subtype of it. Otherwise returns ``false``. This function always succeeds.
```

Reviewer 2: We use imperative style in general so I do not think it is necessary.

   Equivalent to ``PyCMethod_New(ml, self, NULL, NULL)``.

.. c:function:: int PyCFunction_GetFlags(PyObject *func)

>    Get the function flags on *func* as they were passed to

Reviewer: ```suggestion
   Gets the function flags on *func* as they were passed to
```

>    Get the function flags on *func* as they were passed to

Reviewer 2: ```suggestion
   Get the function's flags on *func* as they were passed to
```

   :c:member:`~PyMethodDef.ml_flags`.

   This function returns the function's flags on success, and ``-1`` with an
   exception set on failure.

.. c:function:: int PyCFunction_GET_FLAGS(PyObject *func)

   This is the same as :c:func:`PyCFunction_GetFlags`, but without error
   checking.

.. c:function:: PyCFunction PyCFunction_GetFunction(PyObject *func)

>    Get the function pointer on *func* as it was passed to

Reviewer: ```suggestion
   Gets the function pointer on *func* as it was passed to
```

>    Get the function pointer on *func* as it was passed to

Reviewer 2: Please indicate whether passing a non-function object sets an exception or causes a crash (to make the distinction with `PyCFunction_GET_FUNCTION`)

   :c:member:`~PyMethodDef.ml_meth`.

   This function returns the function pointer on success, and ``NULL`` with an
   exception set on failure.

.. c:function:: int PyCFunction_GET_FUNCTION(PyObject *func)

   This is the same as :c:func:`PyCFunction_GetFunction`, but without error
   checking.

.. c:function:: PyObject *PyCFunction_GetSelf(PyObject *func)

>    Get the "self" object on *func*. This is the object that would be passed

Reviewer: ```suggestion
   Gets the "self" object on *func*. This is the object that would be passed
```

   to the first argument of a :c:type:`PyCFunction`. In modules, this is the

>    module object.

Reviewer 2: "In modules, this is the module object" is unclear. I would either say "For modules"

Author: Let me know what you think of the new wording.

   Equivalent to ``PyCMethod_New(ml, self, NULL, NULL)``.

.. c:function:: int PyCFunction_GetFlags(PyObject *func)

   Get the function's flags on *func* as they were passed to
   :c:member:`~PyMethodDef.ml_flags`.

   If *func* is not a C function object, this fails with a

>    :class:`SystemError`.

Reviewer 2: This one is a regular exception right? so no need to mention it. What I wanted to know is whether passing `func == NULL` produced a SIGSEGV or a Python exception. Since passing NULL is "allowed" (but raises an exception) it's fine to remove the mention to SystemError.

Reviewer 2: My advice is: check other `Check` functions to see how we formulate this. I don't want to be too pedantic but I want to be consistent in general.

Author: > This one is a regular exception right? so no need to mention it. What I wanted to know is whether passing func == NULL produced a SIGSEGV or a Python exception. Since passing NULL is "allowed" (but raises an exception) it's fine to remove the mention to SystemError.

Ah, ok. `NULL` will result in a SIGSEGV, but a non-function will result in a `SystemError`. I updated the text to hopefully clarify that better.

>    *func* must not be ``NULL``.

Reviewer 2: I like this better. Thanks. More generally, maybe we should instead write something at the very beginning of the C API docs about the fact that passing NULL (unless stated otherwise) is an undefined behavior.

Author: We do that already at the top of the "Concrete Objects Layer" section: https://docs.python.org/3/c-api/concrete.html#concrete-objects-layer

> Warning: While the functions described in this chapter carefully check the type of the objects which are passed in, many of them do not check for NULL being passed instead of a valid object. Allowing NULL to be passed in can cause memory access violations and immediate termination of the interpreter.

But, people are forgetful, so I don't mind keeping the note on individual functions.

Reviewer 2: Ok, let's leave it as is!

   This function returns the function's flags on success, and ``-1`` with an
   exception set on failure.

.. c:function:: int PyCFunction_GET_FLAGS(PyObject *func)

   This is the same as :c:func:`PyCFunction_GetFlags`, but without error
   or type checking.

.. c:function:: PyCFunction PyCFunction_GetFunction(PyObject *func)

   Get the function pointer on *func* as it was passed to
   :c:member:`~PyMethodDef.ml_meth`.

   If *func* is not a C function object, this fails with an exception.
   *func* must not be ``NULL``.

   This function returns the function pointer on success, and ``NULL`` with an
   exception set on failure.

> 

Reviewer 2: We have 3 blank lines here. I'll let you edit this locally
